### PR TITLE
Keep track of build re-queuing.

### PR DESCRIPTION
**Pull request by Moo64c**
#210

~~WIP~~
- [x] Field export
- [x] Logic
- [x] Tests.

Ready.

### CODE REVIEW

File: behat/features/ci_build_msg.feature [hunk 1]
```diff
@@ -14,3 +14,20 @@ Feature: CI Build Message
...
+    Given I login with user "William"
+    When  I create repository and CI build "William/app10"
+    And   "1" CI build item for CI build "William/app10" are set to status "in_progress"
+    Then  The "field_requeue_count" value should be "0" for the CI build "William/app10"
+
+  @api
+  Scenario: Check CI Build message requeue counter is 1 after changing from "in_progress" to "queue".
+    Given I login with user "William"
+    When  I create repository and CI build "William/app11"
+    And   "1" CI build item for CI build "William/app11" are set to status "in_progress"
```
Comment (amitaibu) [inline behat/features/ci_build_msg.feature]: make it human readable. Remove the `"1"` and `in_progress` => `In progress`

**Comment by Moo64c** [comment]
@amitaibu Ready for CR.

**Comment by amitaibu** [comment]
- Needs a re-roll
- Fix PR's title
- Bonus points for screenshot :)

**Comment by Moo64c** [comment]
selection_138

File: shoov/modules/custom/shoov_ci_build/shoov_ci_build.module [hunk 2]
```diff
@@ -527,3 +537,25 @@ function shoov_ci_build_get_items_count($node, $status = array()) {
...
 }
+
+/**
+ * Helper function; Check whether a ci_build message was re-queued.
+ *
+ * @param Message $message
+ *   Message from hook_message_presave().
+ */
+function _shoov_ci_build_check_requeue(Message $message) {
+  if (empty($message) || empty($message->original) || $message->type !='ci_build') {
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_ci_build/shoov_ci_build.module]: why `empty($message)` needed?

File: shoov/modules/custom/shoov_ci_build/shoov_ci_build.module [hunk 3]
```diff
@@ -527,3 +537,25 @@ function shoov_ci_build_get_items_count($node, $status = array()) {
...
+    // Not updating a ci_build message.
+    return;
+  }
+
+  $wrapper = entity_metadata_wrapper('message', $message);
+  $old_wrapper = entity_metadata_wrapper('message', $message->original);
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_ci_build/shoov_ci_build.module]: maybe better `$old_wrapper` => `$original_wrapper`

File: behat/features/ci_build_msg.feature [hunk 4]
```diff
@@ -14,3 +14,20 @@ Feature: CI Build Message
     | shoov-tester/gizra-behat branch master Done        |
     | shoov-tester/gizra-behat branch master In progress |
     | Gizra/Gizra branch develop Queue                   |
+
+  @api
+  Scenario: Check CI Build message requeue counter is 0 when created.
+    Given I login with user "William"
+    When  I create repository and CI build "William/app10"
+    And   The CI build item for CI build "William/app10" is set to status "In progress"
+    Then  The "Requeue Count" field value should be "0" for the CI build "William/app10"
```
Comment (amitaibu) [inline behat/features/ci_build_msg.feature]: :+1:

File: behat/features/bootstrap/FeatureContext.php [hunk 5]
```diff
@@ -464,6 +465,104 @@ protected function getNodeByTitleAndBundle($title, $bundle) {
...
+  /**
+   * @When I change CI build :ci_build_title status from :old_status to :new_status
+   */
+  public function iChangeCiBuildStatusFromTo($ci_build_title, $old_status, $new_status) {
+    $old_status = $this->getMachineName($old_status);
+    $new_status = $this->getMachineName($new_status);
+
+    $params = array(
+      '@title' => $ci_build_title,
+      '@old_status' => $old_status,
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php]: old => original

File: behat/features/bootstrap/FeatureContext.php [hunk 6]
```diff
@@ -464,6 +465,104 @@ protected function getNodeByTitleAndBundle($title, $bundle) {
...
+    $result = $query
+      ->entityCondition('entity_type', 'message')
+      ->fieldCondition('field_ci_build', 'target_id', $ci_build_node->nid)
+      ->propertyOrderBy('mid', 'ASC')
+      ->range(0, 1)
+      ->execute();
+
+
+    $wrapper = entity_metadata_wrapper('message', key($result['message']));
+    if ($wrapper->{$field_name}->value() != $field_value) {
```
Comment (amitaibu) [inline behat/features/bootstrap/FeatureContext.php:500]: this part is not clear to me

**Comment by Moo64c** [comment]
@amitaibu For the last part, the test string wasn't clear enough.
Changed to `Then  The "..." field value should be ".." for the last build of CI build "..."`

**Comment by amitaibu** [comment]
> Changed to Then The "..." field value should be ".." for the last build of CI build "..."

? :)

**Comment by Moo64c** [comment]
Just the wording in the behat test..

**Comment by Moo64c** [comment]
@amitaibu 
After pushing, only needed to 
1. Clear cache
2. Feature revert (all - but specifically `shoov_ci_incident` and `shoov_ci_build`
3. Another clear cache for good measure.

**Comment by Moo64c** [comment]
@amitaibu Re-rolled this.

File: shoov/modules/custom/shoov_ci_build/shoov_ci_build.module [hunk 7]
```diff
@@ -542,3 +549,25 @@ function shoov_ci_build_get_items_count($node, $status = array()) {
...
 }
+
+/**
+ * Helper function; Check whether a ci_build message was re-queued.
+ *
+ * @param Message $message
+ *   Message from hook_message_presave().
+ */
+function _shoov_ci_build_check_requeue(Message $message) {
+  if (empty($message->original) || $message->type !='ci_build') {
```
Comment (amitaibu) [inline shoov/modules/custom/shoov_ci_build/shoov_ci_build.module:560]: `empty($message->original)`  => `empty($message->mid)`

Comment (amitaibu) [inline shoov/modules/custom/shoov_ci_build/shoov_ci_build.module:560]: actually, it's fine for now.

**Comment by amitaibu** [comment]
Merged.

**Pull request closed without merging**